OpenRA.Mods.Common/Traits/AutoTarget.cs

```diff
@@ -174,20 +175,19 @@ Actor ChooseTarget(Actor self, WDist range, bool allowMove)
 					var target = Target.FromActor(a);
 					return new KeyValuePair<Armament, Actor>(
 						attack.ChooseArmamentsForTarget(target, false)
-							.Where(arm => allowMove
+							.FirstOrDefault(arm => allowMove
 								|| (target.IsInRange(self.CenterPosition, arm.MaxRange())
-								&& !target.IsInRange(self.CenterPosition, arm.Weapon.MinRange)))
-							.FirstOrDefault(), a);
+								&& !target.IsInRange(self.CenterPosition, arm.Weapon.MinRange))), a);
 				})
 
-				.Where(kv => kv.Key != null && kv.Value.TraitOrDefault<AutoTargetIgnore>() == null)
+				.Where(kv => kv.Key != null && self.Owner.CanViewActor(kv.Value))
```

Comment:
Performance. I wanted to do the visibility check as late as possible, because at that point chances are that most of the candidates have already been excluded.